src/core/ext/filters/http/message_compress/compression_filter.cc

```diff
@@ -0,0 +1,312 @@
+      const size_t after_size = tmp.Length();
+      const float savings_ratio = 1.0f - static_cast<float>(after_size) /
+                                             static_cast<float>(before_size);
+      GPR_ASSERT(grpc_compression_algorithm_name(algorithm, &algo_name));
+      gpr_log(GPR_INFO,
+              "Compressed[%s] %" PRIuPTR " bytes vs. %" PRIuPTR
+              " bytes (%.2f%% savings)",
+              algo_name, before_size, after_size, 100 * savings_ratio);
+    }
+    tmp.Swap(payload);
+    flags |= GRPC_WRITE_INTERNAL_COMPRESS;
+  } else {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_compression_trace)) {
+      const char* algo_name;
+      GPR_ASSERT(grpc_compression_algorithm_name(algorithm, &algo_name));
+      gpr_log(GPR_INFO,
+              "Algorithm '%s' enabled but decided not to compress. Input size: "
+              "%" PRIuPTR,
+              algo_name, payload->Length());
+    }
+  }
+  return message;
+}
+
+absl::StatusOr<MessageHandle> CompressionFilter::DecompressMessage(
+    MessageHandle message, grpc_compression_algorithm algorithm,
+    absl::optional<uint32_t> max_recv_message_length) const {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_compression_trace)) {
+    gpr_log(GPR_ERROR, "DecompressMessage: len=%" PRIdPTR " max=%d alg=%d",
+            message->payload()->Length(), max_recv_message_length.value_or(-1),
+            algorithm);
+  }
+  // Check max message length.
+  if (max_recv_message_length.has_value() &&
+      message->payload()->Length() >
+          static_cast<size_t>(*max_recv_message_length)) {
+    return absl::ResourceExhaustedError(absl::StrFormat(
+        "Received message larger than max (%u vs. %d)",
+        message->payload()->Length(), *max_recv_message_length));
+  }
+  // Check if decompression is enabled (if not, we can just pass the message
+  // up).
+  if (!enable_decompression_ ||
+      (message->flags() & GRPC_WRITE_INTERNAL_COMPRESS) == 0) {
+    return std::move(message);
+  }
+  // Try to decompress the payload.
+  SliceBuffer decompressed_slices;
+  if (grpc_msg_decompress(algorithm, message->payload()->c_slice_buffer(),
+                          decompressed_slices.c_slice_buffer()) == 0) {
+    return absl::InternalError(
+        absl::StrCat("Unexpected error decompressing data for algorithm ",
+                     CompressionAlgorithmAsString(algorithm)));
+  }
+  // Swap the decompressed slices into the message.
+  message->payload()->Swap(&decompressed_slices);
+  message->mutable_flags() &= ~GRPC_WRITE_INTERNAL_COMPRESS;
+  return std::move(message);
+}
+
+class CompressionFilter::DecompressLoop {
+ public:
+  explicit DecompressLoop(CompressionFilter* filter, CallArgs& call_args)
+      : filter_(filter),
+        mapper_(PipeMapper<MessageHandle>::Intercept(
+            *call_args.incoming_messages)) {}
+
+  // Once we have a compression algorithm we can construct the decompression
+  // loop.
+  auto TakeAndRun(grpc_compression_algorithm algorithm) {
```

Comment:
Please add a comment that says something like "this returns a promise that resolves to (whatever)".

Reply:
done